File: swift-bridge/Sources/SpriteKitBridge/EffectNode.swift
import SpriteKit

@_cdecl("sk_effect_node_new")
public func sk_effect_node_new() -> UnsafeMutableRawPointer? {
    skRetain(SKEffectNode())
}

@_cdecl("sk_effect_node_get_should_enable_effects")
public func sk_effect_node_get_should_enable_effects(_ nodeHandle: UnsafeMutableRawPointer?) -> Bool {
    guard let node: SKEffectNode = skBorrow(nodeHandle) else { return false }
    return node.shouldEnableEffects
}

@_cdecl("sk_effect_node_set_should_enable_effects")
public func sk_effect_node_set_should_enable_effects(_ nodeHandle: UnsafeMutableRawPointer?, _ enable: Bool) {
    guard let node: SKEffectNode = skBorrow(nodeHandle) else { return }
    node.shouldEnableEffects = enable
}

@_cdecl("sk_effect_node_get_should_rasterize")
public func sk_effect_node_get_should_rasterize(_ nodeHandle: UnsafeMutableRawPointer?) -> Bool {
    guard let node: SKEffectNode = skBorrow(nodeHandle) else { return false }
    return node.shouldRasterize
}

@_cdecl("sk_effect_node_set_should_rasterize")
public func sk_effect_node_set_should_rasterize(_ nodeHandle: UnsafeMutableRawPointer?, _ rasterize: Bool) {
    guard let node: SKEffectNode = skBorrow(nodeHandle) else { return }
    node.shouldRasterize = rasterize
}

@_cdecl("sk_effect_node_get_blend_mode")
public func sk_effect_node_get_blend_mode(_ nodeHandle: UnsafeMutableRawPointer?) -> Int32 {
    guard let node: SKEffectNode = skBorrow(nodeHandle) else { return 0 }
    return Int32(node.blendMode.rawValue)
}

@_cdecl("sk_effect_node_set_blend_mode")
public func sk_effect_node_set_blend_mode(_ nodeHandle: UnsafeMutableRawPointer?, _ mode: Int32) {
    guard let node: SKEffectNode = skBorrow(nodeHandle) else { return }
    node.blendMode = skBlendMode(mode)
}

File: src/effect_node.rs
use crate::ffi;
use crate::node::AsNode;
use crate::physics::BlendMode;
use crate::private::handle_type;

handle_type!(EffectNode);

impl AsNode for EffectNode {
    fn as_node_ptr(&self) -> *mut core::ffi::c_void {
        self.ptr
    }
}

impl EffectNode {
    #[must_use]
    pub fn new() -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_effect_node_new()) }
    }

    #[must_use]
    pub fn should_enable_effects(&self) -> bool {
        unsafe { ffi::sk_effect_node_get_should_enable_effects(self.ptr) }
    }

    pub fn set_should_enable_effects(&self, enable: bool) {
        unsafe { ffi::sk_effect_node_set_should_enable_effects(self.ptr, enable) };
    }

    #[must_use]
    pub fn should_rasterize(&self) -> bool {
        unsafe { ffi::sk_effect_node_get_should_rasterize(self.ptr) }
    }

    pub fn set_should_rasterize(&self, rasterize: bool) {
        unsafe { ffi::sk_effect_node_set_should_rasterize(self.ptr, rasterize) };
    }

    #[must_use]
    pub fn blend_mode(&self) -> BlendMode {
        BlendMode::from_raw(unsafe { ffi::sk_effect_node_get_blend_mode(self.ptr) })
    }

    pub fn set_blend_mode(&self, mode: BlendMode) {
        unsafe { ffi::sk_effect_node_set_blend_mode(self.ptr, mode as i32) };
    }
}

impl Default for EffectNode {
    fn default() -> Self {
        Self::new().expect("SKEffectNode creation failed")
    }
}
